clippy_lints/src/if_not_else.rs

```diff
@@ -46,13 +47,21 @@ declare_clippy_lint! {
 
 declare_lint_pass!(IfNotElse => [IF_NOT_ELSE]);
 
-impl EarlyLintPass for IfNotElse {
-    fn check_expr(&mut self, cx: &EarlyContext<'_>, item: &Expr) {
-        if in_external_macro(cx.sess, item.span) {
+impl LateLintPass<'_> for IfNotElse {
+    fn check_expr(&mut self, cx: &LateContext<'_>, item: &Expr<'_>) {
+        if in_external_macro(cx.sess(), item.span) {
```

Comment:
This will fix it if my suspicion is right - changing to LateLintPass means that desugaring occurs before the lint runs. `from_expansion` will catch desugaring in addition to macros.

```suggestion
        if item.span.from_expansion() {
```

Reply:
Thanks @camsteffen. This is my first time working with rustc. I didn't know this life saver function existed!
I've been debugging the issue. After dumping the expression trees, I could see that the while loop was indeed represented as an `ExprKind::If` when using `LateLintPass`which caused this problem. I thought about reverting back to `EarlyLintPass` and looking for a workaround, but your comment came just in time!